src/ripple/overlay/impl/ProtocolMessage.h

```diff
@@ -63,33 +61,143 @@ protocolMessageName (int type)
     case protocol::mtGET_OBJECTS:           return "get_objects";
     default:
         break;
-    };
+    }
     return "unknown";
 }
 
 namespace detail {
 
-template <class T, class Buffers, class Handler>
-std::enable_if_t<std::is_base_of<
-    ::google::protobuf::Message, T>::value,
-        boost::system::error_code>
-invoke (int type, Buffers const& buffers,
+struct MessageHeader
+{
+    /** The size of the message on the wire.
+
+        @note This is the sum of sizes of the header and the payload.
+    */
+    std::uint32_t total_wire_size = 0;
+
+    /** The size of the header associated with this message. */
+    std::uint32_t header_size = 0;
+
+    /** The size of the payload on the wire.
+
+        @note If the message is compressed, this value represents the
+              compressed size.
+    */
+    std::uint32_t payload_wire_size = 0;
+
+    /** The size of the payload.
+
+        @note If compression or other message packing techniques are
+              used, this value represents the size of they payload
+              after decompression.
+    */
+    std::uint32_t payload_real_size = 0;
+
+    /** The type of the message. */
+    std::uint16_t message_type = 0;
+
+    /** Indicates whether the message is compressed. */
+    bool compressed = false;
+
+    /** Indicates which compression algorithm the payload is compressed with. */
+    int algorithm = 0;
+};
+
+template <class BufferSequence>
+boost::optional<MessageHeader> parseMessageHeader(
+    BufferSequence const& bufs,
+    std::size_t size)
+{
+    auto iter = boost::asio::buffers_iterator<BufferSequence, std::uint8_t>::begin(bufs);
+
+    MessageHeader hdr;
+
+    // Version 1 header: uncompressed payload.
+    // The top six bits of the first byte are 0.
+    if ((*iter & 0xFC) == 0)
+    {
+        hdr.header_size = 6;
+
+        if (size < hdr.header_size)
+            return {};
+
+        hdr.compressed = false;
+        hdr.algorithm = 0;
+
+        for (int i = 0; i != 4; ++i)
+            hdr.payload_wire_size = (hdr.payload_wire_size << 8) + *iter++;
+
+        hdr.payload_real_size = hdr.payload_wire_size;
+
+        hdr.total_wire_size = hdr.header_size + hdr.payload_wire_size;
+
+        for (int i = 0; i != 2; ++i)
+            hdr.message_type = (hdr.message_type << 8) + *iter++;
+
+        return hdr;
+    }
+
+    // Version 2 header: compressed payload.
```

Comment:
I hear you. I actually don't really like this binary prefix business, and am seriously considering migrating to a pure WebSocket-based transport (which would make the binary header unnecessary) or adopting an approach closer to what you describe (involving coded protobuf streams).

Reply:
gRPC might be worth a look too then...

Reply:
I think that gRPC is probably overkill and will require more rearchitecting than I'd like to do at this time. But I will take a look.

Reply:
I just started looking into libp2p as well (used by Ethereum 2.0, IPFS etc.), but there's no C++ implementation so far and I suspect they are also vulnerable to active MITM with their TLS handshake.

Cap'n'Proto might also be worth a look, gRPC has the advantage that it might also take care of the transport stuff and is probably more widely used and implemented in other software. Wouldn't solve the MITM handshake mitigation, but there were options discussed for that already.

Reply:
Ok, libp2p's handshake is not vulnerable and also prevents MITM attacks with a standard certificate based approach. The only extra part is that one needs to check that the host key embedded in the certificate did actually sign the public key of the certificate (and that the connection is using this key). That's more related to establishing a connection though, not so much to the stream of protobuf messages here.

Reply:
While you can certainly address this issue using certificates, I still think the extra work that users have to go through makes it unlikely that they will. The primary objection to the existing MITM mitigation technique seems to be "yuck, I don't like how this is done." And that's fine but it's not really disposative.

Reply:
No, it is "this is not possible without adding a patch to Python's stdlib" in my case. Accessing these messages is just not possible in a lot of implementations of TLS, verifying a special field in a certificate is.

The certs in libp2p are self-signed and can be created by the application ad-hoc, so there's no work on the user's end...

Reply:
OK. I'll give this some further thought.

Reply:
Hmm. This actually does give me an idea. I may do this, but it'll have to wait for a follow-up PR.